src/openloco/companymgr.cpp

```diff
@@ -61,13 +66,46 @@ namespace openloco::companymgr
             if (_byte_525FCB >= 192)
             {
                 _byte_525FCB = 0;
-                sub_4306D1();
+                produce_companies();
             }
         }
     }
 
-    static void sub_4306D1()
+    static void sub_42F9AC()
+    {
+        call(0x0042F9AC);
+    }
+
+    // 0x004306D1
+    static void produce_companies()
     {
-        call(0x004306D1);
+        if (!_company_competition_delay && _company_max_competing)
+        {
+            uint32_t active = 0;
+            for (auto&& company : companies())
+            {
+                auto id = company.id();
+                if (!company.empty() && id != _byte_525E3C && id != _byte_525E3D)
+                {
+                    active++;
+                }
+            }
+
+            auto& prng = gprng();
+
+            // NOTE: The original code uses srand0 before its mixed.
+            uint32_t rand_val = prng.rand_next();
+            rand_val >>= 8;
+
+            if (!(rand_val & 0xF))
```

Comment:
Okay after looking at IDA I think the best way to do this is to do what we did in OpenRCT2 and convert this into two calls to prng.rand_next(). This one is just a check to only allow the function in 1 out of 16 cases. So

```
if (prng.rand_next(16) == 0)
{
```

The second rand is to get a chance value that is bounded by max_competing.

```
if (prng.rand_next(_company_max_competing) + 1 > active)
{
```